### PR TITLE
[ABW-3108] Factor source access

### PR DESCRIPTION
**Pull request by matiasbzurovski**
Jira ticket: [ABW-3108](https://radixdlt.atlassian.net/browse/ABW-3108?atlOrigin=eyJpIjoiNGY3ZjJlNmQzNTI0NGU5N2E1N2E1MDY2NTUwN2UyZTYiLCJwIjoiaiJ9)

## Description
Currently, the access to factor sources is made in different ways (presented modally, pushed into a navigation, with detail context, as plain white view, etc). This PR adds a new `FactorSourceAccess` component that will be presented anytime we need the user to interact with a factor source. It currently supports `device` and `ledger`, but it can easily be extended to other types in the future (Yubikey, Arculus card, etc).

The component is responsible of showing the correct UI, allowing the user to cancel or retry (if possible). Yet, the logic is still taken care on the same modules of the app (which are `Signing` & `DerivePublicKeys`). Such logic is quite different depending on the context, and putting it altogether in one single component would break encapsulation principles. Also, its complexity and size would make the component very hard to follow.

### Notes
I've added some explanatory comments on the code, including some on deleted classes. Please make sure to expand those files to see those comments 🙂 

## Before & After

<details>
  <summary>Sign transaction with device</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/5556a054-661c-4b1b-97c7-6a98525717cf">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/dad9ecb2-351d-446c-b054-6d215bdc19ef">|
</details>

<details>
  <summary>Sign transaction with Ledger & device</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/394a2841-5f93-4875-8926-42e4ffd8795a">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/d39176c8-9b59-45f9-8516-a9290e51001f">|
</details>

<details>
  <summary>New account with device</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/e29fcd06-11ca-4560-b54b-17ed2515c2e2">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/d09c1d49-421e-4988-bbfe-20a22d0ae4f8">|
</details>

<details>
  <summary>New account with Ledger</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/d165d54f-ac82-4f2a-bb66-e813ba8e5c8f">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/0ff71a76-053b-4091-979c-4d3ea0185093">|
</details>

<details>
  <summary>New Persona</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/766f5cc1-4231-49d4-b84a-8dec881b36d4">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/60b6ab8a-7509-470d-9a9e-eb983e1b8905">|
</details>

<details>
  <summary>Import Olympia Hardware Account with Ledger set up</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/69c66c3a-6153-4273-9a2a-669072c66398">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/8268507c-2171-4c43-a6fc-ffa69ceb13fa">|
</details>

<details>
  <summary>Import Olympia Hardware Account without Ledger set up</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/23519f7e-cedf-44a4-a32d-9aef13911b63">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/7af56c16-b6a1-497a-8f7b-3c9d8df90eb3">|
</details>

<details>
  <summary>ROLA verify babylon accounts</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/870ebb33-ca42-49a9-8b0f-856bc86a916d">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/48f1231e-065c-4725-afb2-c64eb0c3426e">|
</details>

<details>
  <summary>ROLA verify olympia accounts</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/b83d31c3-8073-4b79-8f9b-170cb673e217">  | <video src="https://github.com/radixdlt/babylon-wallet-ios/assets/164921079/e023bfc9-4e38-4b74-9b54-420996f8e4d2">|
</details>

[ABW-3108]: https://radixdlt.atlassian.net/browse/ABW-3108?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: RadixWallet/Features/CreateAccount/Children/CreationOfAccount/CreationOfAccount+Reducer.swift [hunk 1]
```diff
@@ -1,160 +0,0 @@
...
-
-public struct CreationOfAccount: Sendable, FeatureReducer {
-	public struct State: Sendable, Hashable {
-		public let name: NonEmptyString
-		public let networkID: NetworkID?
-		public let isCreatingLedgerAccount: Bool
-		public enum Step: Sendable, Hashable {
-			case step0_chooseLedger(LedgerHardwareDevices.State)
-			case step1_derivePublicKeys(DerivePublicKeys.State)
-		}
```
Comment (matiasbzurovski) [inline RadixWallet/Features/CreateAccount/Children/CreationOfAccount/CreationOfAccount+Reducer.swift:12]: This reducer is deleted since:
- the choose ledger flow can be simply triggered from the coordinator using the `LedgerHardwareDevices` as a step
- the derive public keys flow is no wanted as a child of `CreateAccount` that gets pushed

File: RadixWallet/Features/DerivePublicKeys/DerivePublicKeys+View.swift [hunk 2]
```diff
@@ -1,93 +1,24 @@
...
 // MARK: - DerivePublicKeys.View
 extension DerivePublicKeys {
-	public struct ViewState: Equatable {
-		public let ledger: LedgerHardwareWalletFactorSource?
-		let purpose: DerivePublicKeys.State.Purpose
-
-		var title: String {
-			switch purpose {
-			case .createNewEntity(.account):
-				L10n.DerivePublicKeys.titleCreateAccount
```
Comment (matiasbzurovski) [inline RadixWallet/Features/DerivePublicKeys/DerivePublicKeys+View.swift:22]: A lot of unused strings to be deleted on non-draft version of the PR

File: RadixWallet/Features/DerivePublicKeys/DerivePublicKeys.swift [hunk 3]
```diff
@@ -181,6 +211,22 @@ public struct DerivePublicKeys: Sendable, FeatureReducer {
...
 	}
+
+	public func reduce(into state: inout State, childAction: ChildAction) -> Effect<Action> {
+		switch childAction {
+		case .factorSourceAccess(.delegate(.perform)):
+			.run { send in
+				/// For more information about that `sleep` please  check [this discussion in Slack](https://rdxworks.slack.com/archives/C03QFAWBRNX/p1687967412207119?thread_ts=1687964494.772899&cid=C03QFAWBRNX)
+				@Dependency(\.continuousClock) var clock
+				try? await clock.sleep(for: .milliseconds(700))
+				await send(.internal(.delayedStart))
```
Comment (matiasbzurovski) [inline RadixWallet/Features/DerivePublicKeys/DerivePublicKeys.swift:222]: I am not sure if we need to keep the delay now that the process is moved from a view pushed into a navigation into a sheet modally presented.

File: RadixWallet/Features/Signing/Children/SignWithFactorSource+View.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	struct View: SwiftUI.View {
+		private let store: StoreOf<SignWithFactorSource>
+
+		public init(store: StoreOf<SignWithFactorSource>) {
+			self.store = store
+		}
+
+		public var body: some SwiftUI.View {
+			// We need to create two different views so that each of them triggers the corresponding `.onFirstTask()` that
+			// starts the factor source access. Otherwise, when having multiple signatures chained, it would only trigger the first one.
```
Comment (matiasbzurovski) [inline RadixWallet/Features/Signing/Children/SignWithFactorSource+View.swift:13]: Not a big fan of this approach, but will work on an improvement later. Open to suggestions.

File: RadixWallet/Features/Signing/Coordinator/Signing+View.swift [hunk 5]
```diff
@@ -13,41 +13,7 @@ extension Signing {
...
-			self.store = store
-		}
-
-		public var body: some SwiftUI.View {
-			WithNavigationBar {
-				store.send(.view(.closeButtonTapped))
-			} content: {
-				View(store: store)
-			}
+			SignWithFactorSource.View(store: store.scope(state: \.signWithFactorSource, action: \.child.signWithFactorSource))
```
Comment (matiasbzurovski) [inline RadixWallet/Features/Signing/Coordinator/Signing+View.swift:16]: One could argue that we could merge `Signing` with `SignWithFactorSource`. However, I would rather keep it separately so that each of them is in charge of one task:
- `SignWithFactorSource`: responsible for the signature with the given factor source
- `Signing`: responsible for managing all the factor sources and acting accordingly when all signatures have been collected.

File: RadixWallet/Features/DerivePublicKeys/DerivePublicKeys.swift [hunk 6]
```diff
@@ -63,9 +77,29 @@ public struct DerivePublicKeys: Sendable, FeatureReducer {
...
+					case .device:
+						.device
+					case let .ledger(ledger):
+						.ledger(ledger)
+					default:
+						fatalError("Implement")
+					}
+				case .specificPrivateHDFactorSource:
+					// TODO: MATI: Confirm this should use device
+					.device
```
Comment (CyonAlexRDX) [inline RadixWallet/Features/DerivePublicKeys/DerivePublicKeys.swift:96]: yes that is correct, "private HD" means "A have this mnemonic in hand, I'm gonna use that, and then later save it into Keychain and Profile, and since it is then ON MY PHONE, ON MY DEVICE, it is `Device`-FactorSource" :)

File: RadixWallet/Features/ImportOlympiaLedgerAccountsAndFactorSources/ImportOlympiaLedgerAccountsAndFactorSources.swift [hunk 7]
```diff
@@ -347,122 +353,19 @@ extension LedgerHardwareWalletFactorSource.DeviceModel {
 	}
 }
 
-// MARK: - NameLedgerAndDerivePublicKeys
-public struct NameLedgerAndDerivePublicKeys: Sendable, FeatureReducer {
```
Comment (matiasbzurovski) [inline RadixWallet/Features/ImportOlympiaLedgerAccountsAndFactorSources/ImportOlympiaLedgerAccountsAndFactorSources.swift:351]: This was moved into its own file, and reduced to only take care of the `NameLedger` part. The `DerivePublicKeys` is handled inside here (`ImportOlympiaLedgerAccountsAndFactorSources`).

File: RadixWallet/Features/Signing/Coordinator/Signing+View.swift (hunk 5 above)
Comment (CyonAlexRDX) [reply RadixWallet/Features/Signing/Coordinator/Signing+View.swift:16]: Yes we cannot merge those, the `Signing` is the coordinator and will need to be kept around.

File: RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift [hunk 8]
```diff
@@ -84,7 +98,14 @@ public struct CreatePersonaCoordinator: Sendable, FeatureReducer {
 		case completed
 	}
 
+	public enum InternalAction: Sendable, Equatable {
+		case derivePublicKeys
```
Comment (CyonAlexRDX) [inline RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift:102]: It is in fact a single key, so maybe `derivePublicKey`

File: RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift [hunk 9]
```diff
@@ -61,15 +62,28 @@ public struct CreatePersonaCoordinator: Sendable, FeatureReducer {
...
-			Scope(state: /State.step3_completion, action: /Action.step3_completion) {
+			Scope(state: /State.step2_completion, action: /Action.step2_completion) {
 				NewPersonaCompletion()
 			}
 		}
 	}
 
+	public struct Destination: DestinationReducer {
+		public enum State: Sendable, Hashable {
+			case derivePublicKeys(DerivePublicKeys.State)
```
Comment (CyonAlexRDX) [inline RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift:73]: `derivePublicKey` singular :)

File: RadixWallet/Features/FactorSourceAccess/FactorSourceAccess+View.swift [hunk 10]
```diff
@@ -0,0 +1,82 @@
+// MARK: - FactorSourceAccess.View
+
+public extension FactorSourceAccess {
+	struct ViewState: Equatable {
+		let title: String
+		let message: String
+		let externalDevice: String?
+		let retryEnabled: Bool
```
Comment (CyonAlexRDX) [inline RadixWallet/Features/FactorSourceAccess/FactorSourceAccess+View.swift:8]: we have competing standards, I like prefixing all bools with `is` or `should`, but Gustaf does not. so maybe Ghenadie and or Dan can be tie breakers...

File: RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+View.swift [hunk 11]
```diff
@@ -62,11 +63,11 @@ extension CreateAccountCoordinator {
 							action: CreateAccountCoordinator.Path.Action.step1_nameAccount,
 							then: { NameAccount.View(store: $0) }
 						)
-					case .step2_creationOfAccount:
+					case .step2_selectLedger:
```
[the remark this answers is not in the document]
Comment (matiasbzurovski) [reply RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+View.swift:66]: Agree!

File: RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift (hunk 8 above)
Comment (matiasbzurovski) [reply RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift:102]: changing it here, but shouldn't it be everywhere?

File: RadixWallet/Features/FactorSourceAccess/FactorSourceAccess+View.swift (hunk 10 above)
Comment (matiasbzurovski) [reply RadixWallet/Features/FactorSourceAccess/FactorSourceAccess+View.swift:8]: I actually prefer the `is` prefix as well. Could be helpful to vote a standard among the team and follow it @GhenadieVP @kugel3 @danvleju-rdx

File: RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift (hunk 8 above)
Comment (Sajjon) [reply RadixWallet/Features/CreatePersona/Coordinator/CreatePersonaCoordinator+Reducer.swift:102]: Yes indeed

File: RadixWallet/Core/Resources/Resources/en.lproj/Localizable.strings [hunk 12]
```diff
@@ -1038,3 +1020,15 @@ If you have any “Legacy” Accounts (created on the Olympia network) to import
 "survey_reason_heading" = "What’s the main reason for your score?";
 "survey_reason_fieldHint" = "Let us know...";
 "survey_submitButton" = "Submit Feedback - Thanks!";
+"factorSourceAccess_title_signature" = "Signature Request";
```
Comment (GhenadieVP) [inline RadixWallet/Core/Resources/Resources/en.lproj/Localizable.strings:1023]: Are all of these new strings? if yes then those need to be added first with a pull request on https://github.com/radixdlt/apps-localization-src,
there is also a general guide https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3241246746/Localization+process.

But in short, once that PR is merged, a PR like [this](https://github.com/radixdlt/babylon-wallet-ios/pull/1071/files) will be created on the wallet repo, then swiftgen will need to be used to generate the L10n , merged in main, and after that you do rebase your branch.

File: RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift [hunk 13]
```diff
@@ -47,10 +47,15 @@ public struct NameAccount: Sendable, FeatureReducer {
 			return .none
 
 		case let .confirmNameButtonTapped(sanitizedName):
-			return .send(.delegate(.proceed(
+			return .run { _ in
+				await MainActor.run {
```
Comment (GhenadieVP) [inline RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift:51]: Any reason to not use Focus to resign the responder?

File: RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift [hunk 14]
```diff
@@ -112,33 +145,137 @@ extension CreateAccountCoordinator {
...
+
+		default:
+			return .none
+		}
+	}
+
+	public func reduce(into state: inout State, internalAction: InternalAction) -> Effect<Action> {
+		switch internalAction {
+		case .derivePublicKeyFromDevice:
+			state.destination = .derivePublicKey(
```
Comment (GhenadieVP) [inline RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift:176]: Hmm, I guess you do have internal actions `derivePublicKeyFromDevice` and `derivePublicKeyFromLedger` so you have a single place to set the destination. But I'd say it is unnecessary to have additional actions that are pipped through TCA machinery, rather you can create a function which is called directly.

File: RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift [hunk 15]
```diff
@@ -112,33 +145,137 @@ extension CreateAccountCoordinator {
...
-		case let .path(.element(_, action: .step2_creationOfAccount(.delegate(.createdAccount(newAccount))))):
-			state.path.append(.step3_completion(.init(
-				account: newAccount,
+		case let .createAccountResult(.success(account)):
+			return .run { send in
+				try await accountsClient.saveVirtualAccount(account)
+				await send(.internal(.handleAccountCreated(account)))
+			} catch: { error, send in
+				loggerGlobal.error("Failed to save newly created virtual account into profile: \(error)")
+				await send(.internal(.handleFailure))
```
Comment (GhenadieVP) [inline RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift:209]: could have the same pattern as for `createAccountResult`? instead of separate actions for success and failure.

File: RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift [hunk 16]
```diff
@@ -112,33 +145,137 @@ extension CreateAccountCoordinator {
...
 			return .none
+		}
+	}
+
+	public func reduce(into state: inout State, presentedAction: Destination.Action) -> Effect<Action> {
+		switch presentedAction {
+		case let .derivePublicKey(.delegate(.derivedPublicKeys(hdKeys, factorSourceID, networkID))):
+			guard let hdKey = hdKeys.first else {
+				loggerGlobal.error("Failed to create account expected one single key, got: \(hdKeys.count)")
+				return .send(.internal(.handleFailure))
```
Comment (GhenadieVP) [inline RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift:236]: In general, this seems wasteful, a function should be called directly which does the necessary job, the `internal` actions should be emitted from an Effect.

File: RadixWallet/Features/FactorSourceAccess/FactorSourceAccess.swift [hunk 17]
```diff
@@ -0,0 +1,64 @@
+// MARK: - FactorSourceAccess
+public struct FactorSourceAccess: Sendable, FeatureReducer {
```
Comment (GhenadieVP) [inline RadixWallet/Features/FactorSourceAccess/FactorSourceAccess.swift:2]: Not sure if this is worth having as a Feature, could be just a view.

**Review by GhenadieVP — CHANGES_REQUESTED** [review]
Awesome work 👏!
I only have some comments around using Internal actions, as well that strings need to first be added separately.

File: RadixWallet/Core/Resources/Resources/en.lproj/Localizable.strings (hunk 12 above)
Comment (matiasbzurovski) [reply RadixWallet/Core/Resources/Resources/en.lproj/Localizable.strings:1023]: thanks for sharing, I didn't know we had this external `apps-localization-src`. In this case, there are some new Strings (content changed) and some which had the key used renamed. Do I need to sync with Android for this?

File: RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift (hunk 13 above)
Comment (matiasbzurovski) [reply RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift:51]: Seems to me it adds a lot of boilerplate (need to set the focus on the textfield, have the corresponding variable on the state and then also the action handling). It would be useful if we wanted to set the focus when opening the view or something, but if it is just to loose the focus I think this way is simpler.

File: RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift (hunk 15 above)
Comment (matiasbzurovski) [reply RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift:209]: Not following you in this case. You would want to manually send `await send(.internal(.createAccountResult(.failure(error))))` instead?

File: RadixWallet/Features/FactorSourceAccess/FactorSourceAccess.swift (hunk 17 above)
Comment (matiasbzurovski) [reply RadixWallet/Features/FactorSourceAccess/FactorSourceAccess.swift:2]: Agree. However, we did briefly discuss with @CyonAlexRDX about the possibility of moving all the code from `DerivePublicKeys` and `SignWithFactorSource` into this new `FactorSourceAccess`. Not 100% sure this would be a better idea, but if we go on with it then it would definitely be a Feature.

File: RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift [hunk 18]
```diff
@@ -112,37 +143,134 @@ extension CreateAccountCoordinator {
...
+		case .derivePublicKey(.delegate(.cancel)):
+			state.destination = nil
+			return .none
+
 		default:
 			return .none
 		}
 	}
+
+	private func derivePublicKey(state: inout State, ledger: LedgerHardwareWalletFactorSource?) -> Effect<Action> {
```
Comment (danvleju-rdx) [inline RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift:255]: Why not pass `factorSourceOption` directly as a param instead of an optional `ledger`?

File: RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift (hunk 15 above)
Comment (GhenadieVP) [reply RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift:179]: I mean for consistency with `createAccountResult`  instead of having in Internal action enum:
```
case handleAccountCreated(Profile.Network.Account)
case handleFailure
```
you could have:
```
case handleAccountCreated(TaskResult<Profile.Network.Account>)
```

then here you could have:
```
return .run { send in
  let result = TaskResult { try await accountsClient.saveVirtualAccount(account) }
  await send(.internal(.handleAccountCreated(result)))
}
```

File: RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift (hunk 13 above)
Comment (GhenadieVP) [reply RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift:51]: Agree, was just curious.
Could we then extract this in some separate utility that could be use to resign the responder? idk if it could be an extension on `Effect`

File: RadixWallet/Core/Resources/Resources/en.lproj/Localizable.strings (hunk 12 above)
Comment (GhenadieVP) [reply RadixWallet/Core/Resources/Resources/en.lproj/Localizable.strings:1023]: Yes, as they will need to pull in the new strings and update their codebase.

File: RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift (hunk 13 above)
Comment (kugel3) [reply RadixWallet/Features/CreateAccount/Children/NameAccount/NameAccount+Reducer.swift:51]: Not sure about the context here don't don't be afraid to work outside TCA for things like focus, if you don't actually need to control it through a reducer. For UI only things at least, we are not TCA maximalists.

**Review by GhenadieVP — APPROVED** [review]
